### DISCUSSION TITLE
Add more variables in v4

### DISCUSSION THREAD

**Discussion by Hardexit** [PSAppDeployToolkit/PSAppDeployToolkit#1103] [Q&A] [answered]
Hi, I'm trying to get to grips with v4 and add more variables as I need them for INtune. Unfortunately I always get a 60008 error:

![image](https://github.com/user-attachments/assets/bc7813c3-2623-4c2a-b20d-762f3e967eda)

**Comment by mjr4077au** [comment, marked as answer]
Hi @Hardexit, that hashtable can't be extended like that as it's used to instantiate the session, however you can define your extra variables underneath it that you need locally within your deployment script.

**Reply by Hardexit** [reply]
I'm almost sure of that, of course it's not quite as nice when it's splintered. Thanks for the confirmation.

**Comment by Hardexit** [comment]
Hello @mjr4077au, I have now added the variables but unfortunately I have encountered another problem. For the packaging for intune I use the variables from the script, but unfortunately I now get an error message when I do Import-Module Invoke-AppDeployToolkit.ps1:
```
Import-Module: ...\Invoke-AppDeployToolkit.ps1:242
Line |
 242 |      Import-Module -FullyQualifiedName @{ ModuleName = $moduleName; Gu …
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | The specified module 'PSAppDeployToolkit' with version '4.0.3' was not loaded because no valid module file was
     | found in any module directory.
```

Unfortunately, I can't think of an alternative and maintaining all variables twice is not a good option either.

**Reply by DanGough** [reply]
Invoke-AppDeployToolkit.ps1 is not a module, so a literal command `Import-Module Invoke-AppDeployToolkit.ps1` would not work.

The import happens on line 283 of the template Invoke-AppDeployToolkit.ps1, if you're getting an error on 242 perhaps some important code was removed? Post your script and we can probably point out what's wrong.

**Reply by Hardexit** [reply]
The script is a little longer because it has to fulfill some tasks: https://gist.github.com/Hardexit/1a7ce9ea446907f2ad5391c20b74877b#file-functions-psm1-L104

**Reply by Hardexit** [reply]
My best idea at the moment is the following function:
```powershell
function Get-VariableFromScript {
  param (
      [string]$filePath,
      [string]$variableName
  )

  $content = Get-Content -Path $filePath

  $pattern = '^.*(?=\[String\]\$' + [regex]::escape(${variableName}) + '\s*=\s*")'

  $variableLine = ${content} | Where-Object { $_ -match ${pattern} }

  if (${variableLine} -ne '""') {
      $variableValue = (${variableLine} -replace '.*=\s*"([^"]*)".*', '$1').Trim()
      return ${variableValue}
  } else {
      Write-Host "Variable '${variableName}' nicht gefunden in der Datei."
      exit 1
  }
}
```

**Reply by DanGough** [reply]
I'm saying the odd thing is your Invoke-AppDeployToolkit.ps1 script is shorter than the blank template.

This functions.psm1 file is seemingly unrelated to the error you posted previously. But there are a couple of issues with it anhow:

- The variable declarations at the start may not be accessible anywhere
- You may instead want to try adding your custom functions to the template PSAppDeployToolkit.Extensions.psm1, as this is all pre-configured to import automatically
- You can't run Import-Module on a deployment script, as it's not a module:

Line 104: Import-Module ${FilePs1} 2>&1
Line 186: Import-Module ${FolderAPP}\Deploy-Application.ps1 2>&1

**Reply by Hardexit** [reply]
I thought you wanted to know why I call the script, here is the applicationsfile https://gist.github.com/Hardexit/2109452804e29365a1887bc622e87a91

> You can't run Import-Module on a deployment script, as it's not a module:

For not being able to do that, I have been working with it for over a year ;)

**Reply by DanGough** [reply]
Back to the `The specified module 'PSAppDeployToolkit' with version '4.0.3' was not loaded because no valid module file was found in any module directory.` error... Do you either have the module installed from the gallery, or present in `PSAppDeployToolkit\PSAppDeployToolkit.psd1` next to Invoke-AppDeployToolkit.ps1?

**Reply by mjr4077au** [reply]
> My best idea at the moment is the following function:
> 
> ```powershell
> function Get-VariableFromScript {
>   param (
>       [string]$filePath,
>       [string]$variableName
>   )
> 
>   $content = Get-Content -Path $filePath
> 
>   $pattern = '^.*(?=\[String\]\$' + [regex]::escape(${variableName}) + '\s*=\s*")'
> 
>   $variableLine = ${content} | Where-Object { $_ -match ${pattern} }
> 
>   if (${variableLine} -ne '""') {
>       $variableValue = (${variableLine} -replace '.*=\s*"([^"]*)".*', '$1').Trim()
>       return ${variableValue}
>   } else {
>       Write-Host "Variable '${variableName}' nicht gefunden in der Datei."
>       exit 1
>   }
> }
> ```

It's worth mentioning also that you can add these variables to $adtSession now, but only in the upcoming point release: https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/commit/380dfc9e2e8c9c28cd060fceaeeb9b41af34fc4b

**Reply by Hardexit** [reply]
> Do you either have the module installed from the gallery, or present in `PSAppDeployToolkit\PSAppDeployToolkit.psd1` next to Invoke-AppDeployToolkit.ps1?

All files are located in a _template folder in the root directory, in the app itself there is only the required data, i.e. usually a program installation file and the `Invoke-AppDeployToolkit.ps1` nothing more.

Currently my script works so that I only put the changed data in the folder of the appand the script then assembles the entire package by merging the current data with that of the template, then it packs everything into an intune file and uploads it afterwards.

This saves me a lot of memory in my git repo. It also makes updating easier because I don't have to replace files in every app.

I have also thought about whether it would make sense to put all the variables in a config file, but then it would be difficult to use them in the `Invoke-AppDeployToolkit.ps1`.

> 380dfc9

that's great, thank you
